minimesos/src/main/java/com/containersol/minimesos/docker/DockerClientFactory.java

```diff
@@ -3,28 +3,84 @@
+import java.util.Properties;
+
+import static com.github.dockerjava.core.DockerClientConfig.*;
 
 /**
  * Factory for creating {@link DockerClient}s
  */
 public class DockerClientFactory {
-
     private static DockerClient dockerClient;
+    private static DockerClientConfigBuilder configBuilder = new DockerClientConfigBuilder();
+    private static String dockerHostEnv;
+    private static String dockerCertPathEnv;
+    private static String dockerHostPrp;
+    private static String dockerCertPathPrp;
+    private static String dockerUri;
+    private static String dockerCertPath;
+    private static Properties properties = System.getProperties();
+    private static Map<String,String> environment = System.getenv();
 
     public static DockerClient build() {
-        if (dockerClient == null) {
-            DockerClientConfig.DockerClientConfigBuilder builder = DockerClientConfig.createDefaultConfigBuilder();
+        DockerClientConfig config = populateConfigBuilder(configBuilder).build();
+        dockerClient = DockerClientBuilder.getInstance(config).build();
+        return dockerClient;
+    }
 
-            String dockerHostEnv = System.getenv("DOCKER_HOST");
-            if (StringUtils.isBlank(dockerHostEnv)) {
-                builder.withUri("unix:///var/run/docker.sock");
-            }
+    public static DockerClient build(Properties argProperties) {
+        properties = argProperties;
+        return build();
+    }
+
+    public static DockerClient build(Map<String,String> env) {
+        environment = env;
+        return build();
+    }
 
-            DockerClientConfig config = builder.build();
-            dockerClient = DockerClientBuilder.getInstance(config).build();
+    public static DockerClientConfigBuilder populateConfigBuilder(DockerClientConfigBuilder builder) {
+        builder.withVersion("");
+        builder.withDockerCertPath("");
+
+        dockerHostEnv = environment.get("DOCKER_HOST");
+        dockerCertPathEnv = environment.get("DOCKER_CERT_PATH");
+        dockerHostPrp = properties.getProperty("docker.io.url");
+        dockerCertPathPrp = properties.getProperty("docker.io.dockerCertPath");
+        dockerUri = "unix:///var/run/docker.sock";
+        
+        if (dockerHostEnv != null) {
+            dockerUri = dockerHostEnv;
+        } else if (dockerHostPrp != null) {
+            dockerUri = dockerHostPrp;
+            builder.withProperties(properties);
+        }
+        if (dockerCertPathPrp != null && new File(dockerCertPathPrp).exists()) {
+            dockerCertPath = dockerCertPathPrp;
+        }
+        if (dockerCertPathEnv != null && new File(dockerCertPathEnv).exists()) {
+            dockerCertPath = dockerCertPathEnv;
+        }
+        builder.withUri(dockerUri.replace("tcp", "https"));
+        if (dockerCertPath != null) {
+            builder.withDockerCertPath(dockerCertPath);
+        }
+        return builder;
+    }
+
+    public static DockerClient getDockerClient() {
```

Comment:
What is the difference between this method and `build()`? Could you JavaDoc all public methods?